alicloud/data_source_alicloud_vpcs.go

```diff
@@ -0,0 +1,202 @@
+package alicloud
+
+import (
+	"fmt"
+	"github.com/denverdino/aliyungo/ecs"
+	"github.com/hashicorp/terraform/helper/schema"
+	"log"
+	"regexp"
+)
+
+func dataSourceAlicloudVpcs() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceAlicloudVpcsRead,
+
+		Schema: map[string]*schema.Schema{
+			"cidr_block": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"status": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"name_regex": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateNameRegex,
+				ForceNew:     true,
+			},
+			"is_default": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				ForceNew: true,
+			},
+			"vswitch_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			// Computed values
+			"vpcs": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"region_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"status": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"vpc_name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"vswitch_ids": {
+							Type:     schema.TypeList,
+							Computed: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+						"cidr_block": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"vrouter_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"description": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"is_default": {
+							Type:     schema.TypeBool,
+							Computed: true,
+						},
+						"creation_time": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+func dataSourceAlicloudVpcsRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AliyunClient).ecsconn
+
+	cidrBlock, cidrBlockOk := d.GetOk("cidr_block")
+	status, statusOk := d.GetOk("status")
+	nameRegex, nameRegexOk := d.GetOk("name_regex")
+	isDefault, isDefaultOk := d.GetOk("is_default")
+	vswitchId, vswitchIdOk := d.GetOk("vswitch_id")
+
+	args := &ecs.DescribeVpcsArgs{
+		RegionId: getRegion(d, meta),
+	}
+
+	var allVpcs []ecs.VpcSetType
+
+	for {
+		vpcs, paginationResult, err := conn.DescribeVpcs(args)
+		if err != nil {
+			break
+		}
+
+		allVpcs = append(allVpcs, vpcs...)
+
+		pagination := paginationResult.NextPage()
+		if pagination == nil {
+			break
+		}
+
+		args.Pagination = *pagination
+	}
+
+	var filteredVpcsTemp []ecs.VpcSetType
+
+	for _, vpc := range allVpcs {
+		if cidrBlockOk && vpc.CidrBlock != cidrBlock.(string) {
```

Comment:
Line 129 can be simplified by the following code:

if cidrBlock, ok := d.GetOk("cidr_block"); ok && vpc.CidrBlock != cidrBlock.(string) {}

And same as the next lines.